Solution

SAKorney opened the pull request:
Решение = немного parallelStream + CompletableFuture
Кроме подсчёта частоты слов, остальные задачи I/O bound.

SAKorney committed:
First iteration

SAKorney committed:
Refactoring

SAKorney committed:
'Fix'

kincajou commented on parallelism/src/main/java/ru/hh/school/homework/Launcher.java:
```diff
@@ -37,9 +49,86 @@ public static void main(String[] args) throws IOException {
+      return Map.entry(path, calcStat(path, executor));
+    }
+    catch (IOException ignored) {
+      return Map.entry(path, List.of());
```
не очень хорошо, случится exception - ты даже не узнаешь про него

SAKorney replied:
Исправил. Вообще, похоже, в Java обработка исключений (из-за checked exceptions) дело не самое тривиальное. Почувствовал это при использовании Stream и элементов ФП.

kincajou commented on parallelism/src/main/java/ru/hh/school/homework/Launcher.java:
```diff
@@ -37,9 +49,86 @@ public static void main(String[] args) throws IOException {
+    }
+  }
+
+  private static Map.Entry<Path, List<String>> processDirectory(Path path, ExecutorService executor) {
```
можно было бы рекорд коротенький сделать вместо Map.Entry

SAKorney replied:
Сделал. Изначально была похожая идея, просто хотелось ещё пообвыкнуть со стандартной библиотекой

kincajou commented on parallelism/src/main/java/ru/hh/school/homework/Launcher.java:
```diff
@@ -37,9 +49,86 @@ public static void main(String[] args) throws IOException {
+  private static void launch(Path path,  Function<String, Long> search, ExecutorService executor) throws IOException {
+    try (Stream<Path> paths = Files.walk(path)) {
+      paths
+              .parallel()
```
вот тут есть что улучшить
ограниченное число тредов в system fork join pool будут постоянно заняты тем что либо ждут join фьючи обхода дерева, либо join фьючи поиска
можно ходить по дереву и одновременно поиском заниматься

SAKorney replied:
Реализовал. Есть ощущение, что я переусложнил

kincajou commented on parallelism/src/main/java/ru/hh/school/homework/Launcher.java:
```diff
@@ -68,16 +157,31 @@ private static void testSearch() throws IOException {
   }
 
+  private static final Random rnd = new Random();
+  private static long mockSearch(String query) {
```
что naiveSearch что mockSearch можно сильно ускорить если кешировать значения

SAKorney replied:
Добавил кеширование

SAKorney committed:
Improvements

Разные пулы для чтения с диска и для запросов в сеть
Добавлено кеширование результатов запроса
Методы поиска вынесены в отдельные классы

SAKorney committed:
Refactoring

kincajou commented on parallelism/src/main/java/ru/hh/school/homework/Launcher.java:
```diff
@@ -37,9 +49,107 @@ public static void main(String[] args) throws IOException {
+              .supplyAsync(() -> processDirectory(path, executorFileIO))
+              .thenAccept(dir -> processQueries(dir, search, executorNetworkIO))
+              .get();
+    } catch ( InterruptedException | ExecutionException e) {
```
поймал Interrupted - взведи флаг

kincajou commented on parallelism/src/main/java/ru/hh/school/homework/Launcher.java:
```diff
@@ -37,9 +49,107 @@ public static void main(String[] args) throws IOException {
+                              ExecutorService executorNetworkIO) {
+    try {
+      CompletableFuture
+              .supplyAsync(() -> processDirectory(path, executorFileIO))
```
проблема все еще есть
тут написано что:
- в системногом fjp фильтруем директории [fjp тред 1]
- все еще в треде [fjp тред 1] берем еще один тред из системного fjp [fjp тред 2] и запускаем processDirectory внутри которого последовательно обрабатываем файлы (последовательно потому что async а потом сразу join)
- после этого, все еще в треде [fjp тред 2] запускается processQueries внутри которого опять последовательно ищем (опять async и join)
- на выходе из processQueries отпускается [fjp тред 2]
- после того как вся директория обработалась таким образом отпускается [fjp тред 1] - .get()

вощем очень запутано и точно не паралельно

kincajou replied:
грубо, как это должно работать:
для каждой папки -> supplyAsync(считаем топ, пул1).thenApplyAsync(ищем, пул2).thenAccept(выводим в консоль)
фьючи собираем в лист
внутри методов никаких джоинов
в конце allOf(фьючи).join() - как отпустит - все доработало, можно гасить пулы

kincajou replied:
поправлюсь, внутри processDirectory  и processQueries - паралельность есть, т.к. через терминальную toList оно проходит, но все будет все равно в fjp тупить

SAKorney committed:
One more improvement
Вызов *Async методов вынесен в process
Вывод на консоль теперь отдельная операция

kincajou commented on parallelism/src/main/java/ru/hh/school/homework/Launcher.java:
```diff
@@ -37,9 +49,93 @@ public static void main(String[] args) throws IOException {
-    testSearch();
+    int numOfThreads = Runtime.getRuntime().availableProcessors();
+    // Два отдельных пула для чтения с диска и для запросов в сеть
+    ExecutorService executorFileIO = Executors.newFixedThreadPool(numOfThreads/2);
```
в принципе можно не делить на 2, можно вообще поставить на сколько памяти хватит
и те и те задачи - IO bound, за цпу не дерутся

kincajou replied:
но ограничивать само собой хоть чем-то нужно, чтобы out of memory случайно не словить

kincajou commented:
ок, считаем что закончили

мок серч - гуд, и то что поиск для этого абстрагировал - гуд 
с основой паралельности для этой задачки разобрались
пулы останавливаются когда надо

для задач по подсчету слов cf - норм. Идеально сделать конечно на fork join pool, либо написать на паралел стримах с flatmap при подсчете слов в файле с последующим подсчетом топ10 для папки через collect(groupingBy...)) - подсчет хорошо делится и подходит fjp
для задач по поиску - cf подходит лучше чем fjp

при желании можно написать оба варианта и померять через jmh

надеюсь было полезно!
